dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/BinaryEncoding.java

```diff
@@ -81,17 +84,40 @@ public String toString() {
         }
     }
 
-    enum InputStreamDeserializer implements Deserializer<Object> {
+    enum InputStreamDeserializer implements Deserializer<InputStream> {
         INSTANCE;
 
         @Override
-        public Object deserialize(InputStream input) {
-            return input;
+        public InputStream deserialize(InputStream input, Closeable response) {
+            // intentionally not closing just yet, otherwise users wouldn't be able to get any data out of it!
+            return new ResponseClosingInputStream(input, response);
         }
 
         @Override
         public String toString() {
             return "InputStreamDeserializer{}";
         }
     }
+
+    static class ResponseClosingInputStream extends ForwardingInputStream {
+        private final InputStream inputStream;
+        private final Closeable response;
+
+        ResponseClosingInputStream(InputStream inputStream, Closeable response) {
+            this.inputStream = inputStream;
+            this.response = response;
```

Comment:
Not sure it's necessary to pass along the response object here -- all our implementations only require the body stream to be closed, and allow Response.close as a convenience to close the body.

Reply:
Yeah so this was something I wasn't quite sure about - given that people are free to implement their own `Response`, it seemed like there's a possibility that closing the Response and closing the InputStream could release different resources somehow...

Reply:
Thoughts on updating the Response javadoc to note that at least one of response or response.body must be closed, but both aren't required? That way we force implementors to handle resource release -- all our current implementations only hold resources in the body, so I don't think there's risk, and it keeps the api a touch cleaner.

Reply:
Have added a bit of extra wording - does this capture your intentions?

Reply:
:+1: yep, thanks.